Network interface userspace access

jukkar opened the pull request:
Add new functions that can be used to add/lookup/remove IP address from the network interface when the application is running in userspace context. These functions are only available in userspace if CONFIG_NET_IF_USERSPACE_ACCESS is enabled.

jukkar committed:
net: if: Add syscall interface to IP address add and rm

Make IPv4 and IPv6 address addition and removal possible from
userspace app. But allow this only if CONFIG_NET_IF_USERSPACE_ACCESS
By default these operations are not allowed from userspace app.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
tests: net: if: Test userspace support in net_if

Add more tests to test userspace enabled APIs.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: if: Make IPv4 address const in removal function

As the function does not touch IPv4 address in removal, we
can mark it const.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

pfalcon commented:
Inconsistent naming of some functions is pointed by CI.

Beyond that, I'd suggest to explicitly mention in docstrings that interface index is 1-based (that's what we standardized on, right?). And even beyond that, I wonder if we should model such functions based on Linux' API (which would be based on classical BSD Unix API, even if this API is not mandated by POSIX).

jukkar commented:
> I wonder if we should model such functions based on Linux' API (which would be based on classical BSD Unix API, even if this API is not mandated by POSIX).

What Linux APIs you are referring to?

pfalcon commented:
> And even beyond that, I wonder if we should model such functions based on Linux' API (which would be based on classical BSD Unix API

Well, d'oh, that appears to be rather bloated and not directly fitting Zephyr model:

http://man7.org/linux/man-pages/man3/getifaddrs.3.html
http://man7.org/linux/man-pages/man3/if_nametoindex.3.html

Well, I'd still spend more time looking if something can be made "more consistent" with Linux ;-).

jukkar commented:
The functions here can be used as a base if we want to add some BSD versions later.

tbursztyka commented on subsys/net/ip/Kconfig:
```diff
@@ -529,6 +529,22 @@ config NET_DEFAULT_IF_CANBUS
 
 endchoice
 
+config NET_IF_USERSPACE_ADDR_ADD
```
Why having such atomic options? Couldn't we have a unique generic NET_IF_USERSPACE  instead?
If we expose more to the userspace following the same pattern, we will end up with a long list of Kconfig options.

jukkar replied:
I was thinking a bit more fine grained configuration options, but we can of course go with wider options too. I have no big preference here.

jukkar commented:
Updated according to comments.

jukkar commented:
recheck

jukkar commented:
Rebased against latest upstream.

jukkar committed:
net: if: Add userspace support to IPv4 netmask set function

Allow application to call net_if_ipv4_set_netmask_by_index()
and set the netmask if enabled by configuration.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: if: Add userspace support to IPv4 gateway set function

Allow application to call net_if_ipv4_set_gw_by_index()
and set the gateway if enabled by configuration.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: if: Make gateway address in net_if_ipv4_set_gw() const

As the function does not modify the parameter, we can make it const.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
* fixed unit tests
* added IPv4 netmask and gateway set functions so that they can be called from userspace

jukkar commented:
Rebased against latest master.
